java/ql/src/semmle/code/java/frameworks/javase/URL.qll

```diff
@@ -0,0 +1,56 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/* Am URL constructor expression */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URL" }
+
+  Expr hostArg() {
+    // URL(String spec)
+    if this.getNumArgument() = 1
+    then result = this.getArgument(0)
+    else
+      // URL(String protocol, String host, int port, String file)
+      if this.getNumArgument() = 4
+      then result = this.getArgument(1)
+      else
+        // URL(String protocol, String host, int port, String file, URLStreamHandler handler)
+        if this.getNumArgument() = 5
+        then result = this.getArgument(1)
+        else
+          // URL(String protocol, String host, String file)
+          // but not
+          // URL(URL context, String spec, URLStreamHandler handler)
+          if
+            this.getNumArgument() = 3 and
+            not this.getConstructor().getParameter(2).getType().getName() = "URLStreamHandler"
+          then result = this.getArgument(1)
+          else none()
+    // URL(URL context, String spec)
+    // or if this.getNumArgument() = 4
+    // then result = this.getArgument(1)
+  }
+
+  Expr protocolArg() {
+    // In all cases except where the first parameter is a URL, the argument
+    // containing the protocol is the first one, otherwise it is the second.
+    if this.getConstructor().getParameter(0).getType().getName() = "URL"
```

Comment:
Can you not do `if this.getConstructor().getParameter(0).getType() instanceof TypeUrl` instead?

Reply:
That's not my code. I refactor it from `java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql`. 
I will change it if @aschackmull is fine with that change though.